iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,125 @@
+
+    def escape_attr(self, s):
+        return re.sub(r'([<>"&])', lambda m: self.ENTITIES[m.group(0)], s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_text(self, s, escape_mode):
+        if s is None:
+            return ''
+        if escape_mode == EscapeMode.STYLE:
+            # Don't escape >
+            # & and < may only appear in a string or comment in CSS, so escape
+            # as if they're in a string.
+            return re.sub(r'([<&])', lambda m: "\\%06X" % ord(m.group(1)), s)
+        return re.sub(r'([<>&])', lambda m: self.ENTITIES[m.group(0)], s)
+
+    def attributes(self, node):
+        return sorted(
+            '%s="%s"' % (self.tag_as_qname(k, node.nsmap), self.escape_attr(v))
+                for k, v in node.items()
+            )
+
+    def write_comment(self, n, escape_mode):
+        return '<!--' + n.text + '-->' + self.escape_text(n.tail, escape_mode)
+
+    def write_processing_instruction(self, n, escape_mode):
+        return '<?' + n.target + ' ' + n.text + '?>' + self.escape_text(n.tail, escape_mode)
+
+    def write_element(self, n, parent_nsmap = {}, escape_mode = EscapeMode.DEFAULT):
+        name = self.tag_as_qname(n.tag, n.nsmap)
+        qname = etree.QName(n.tag)
+        selfclose = len(n) == 0 and n.text is None and self.is_selfclosable(n)
+        parts = [ name ] + self.namespace_declarations(n.nsmap, parent_nsmap) + self.attributes(n)
+        s = '<' + ' '.join(parts)
+
+        if qname.localname.lower() == 'style':
+            inner_escape_mode = EscapeMode.STYLE
+        else:
+            inner_escape_mode = escape_mode
+
+        if selfclose:
+            s += '/>'
+        else:
+            s += '>'
+            s += self.escape_text(n.text, inner_escape_mode)
+
+            for child in n.iterchildren():
+                if isinstance(child, etree._Comment):
+                    s += self.write_comment(child, inner_escape_mode)
+                elif isinstance(child, etree._ProcessingInstruction):
+                    s += self.write_processing_instruction(child, inner_escape_mode)
+                else:
+                    s += self.write_element(child, n.nsmap, inner_escape_mode)
+            s += '</%s>' % name
+
+        s += self.escape_text(n.tail, escape_mode)
+
+        return s
+
+    def serialize(self, element, fout):
+        if hasattr(element, 'getroot'):
+            element = element.getroot()
+        if self.xml_declaration:
+            fout.write('<?xml version="1.0" encoding="utf-8"?>\n'.encode('utf-8'))
```

Comment:
Out of curiosity, does iXBRL tolerate 1.1?

Can/should we include `standalone="..."` if present in the source?

Reply:
> Out of curiosity, does iXBRL tolerate 1.1?

I don't recall any discussion of this, and the iXBRL spec references XML 1.0, so I think probably not.  That said, it's easy enough to make this code parrot the version number of the input document.  

It doesn't look like lxml supports 1.1, in that it still rejects characters that are now allowed, but I've implemented the new 1.1 requirement that certain control characters are escaped.

> Can/should we include standalone="..." if present in the source?

I see no reason not to repeat include it if the source document specifies it, so I've made it do that.